### DISCUSSION TITLE
Ripgrep fails without meaningful errors when spawned by a node process

### DISCUSSION THREAD

**Discussion by lramos15** [BurntSushi/ripgrep#2047] [Q&A] [answered]
Using `spawn`, `exec`, `execSync`, or `spawnSync` within node ripgrep always seems to exit with exitcode 0. 

Here's a simple node script as an example
```
const cp = require('child_process');

try {
  const retrieved_paths = cp.execSync("rg --debug 'const'", { encoding: 'utf-8', cwd: '/Users/lramos15/vscode'});
  console.log(retrieved_paths);
} catch (err) {
  console.error(err);
}

```

The only `rg` script that will run is `rg --help`. 

The debug info also isn't helpful as it just includes what appears a generic unrelated message
```
    'DEBUG|grep_regex::literal|crates/regex/src/literal.rs:58: literal prefixes detected: Literals { lits: [Complete(const)], limit_size: 250, limit_class: 10 }\n' +
      'DEBUG|globset|crates/globset/src/lib.rs:421: built glob set; 0 literals, 0 basenames, 12 extensions, 0 prefixes, 0 suffixes, 0 required extensions, 0 regexes\n'
```

Any help would be appreciated 🙏

**Comment by lramos15** [comment, marked as answer]
Update:

For anyone else stuck in the same spot I think it has to do with the way child processes and parent process stdin/stdout/stderr interact
explicitly stating `stdio: ['ignore', 'pipe', 'pipe']` fixed the problem

**Reply by BurntSushi** [reply]
Correct. When you don't give a path to ripgrep, it has to _guess_ whether to search stdin or to search the current working directory. If stdin looks like it's readable, then that's what ripgrep will search.

Therefore, an alternative is to use `rg foo ./` instead of `rg foo`. In that case, no guessing is done.

**Reply by lramos15** [reply]
Must of missed this in the documentation. Thank you this also works. I switched to passing in the absolute path instead of setting a `cwd` and using `stdio: ['ignore', 'pipe', 'pipe']` as I feel like this is more clear and doesn't rely on a specific working directory.

**Reply by BurntSushi** [reply]
Agreed. :-)
